pkg/phlaredb/downsample/downsample.go

```diff
@@ -0,0 +1,252 @@
+package downsample
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+
+	"github.com/google/uuid"
+	"github.com/parquet-go/parquet-go"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/prometheus/common/model"
+
+	"github.com/grafana/pyroscope/pkg/phlaredb/block"
+	schemav1 "github.com/grafana/pyroscope/pkg/phlaredb/schemas/v1"
+	"github.com/grafana/pyroscope/pkg/util/build"
+)
+
+type interval struct {
+	durationSeconds int64
+	shortName       string
+}
+
+var (
+	intervals = []interval{
+		{
+			durationSeconds: 5 * 60,
+			shortName:       "5m",
+		},
+		{
+			durationSeconds: 60 * 60,
+			shortName:       "1h",
+		},
+	}
+	inputSamplesHistogram = promauto.NewHistogram(
+		prometheus.HistogramOpts{
+			Name:    "pyroscope_downsampler_input_profile_samples",
+			Help:    "The number of samples per profile before downsampling",
+			Buckets: prometheus.ExponentialBuckets(32, 2, 15),
+		})
+	outputSamplesHistogram = promauto.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Name:    "pyroscope_downsampler_output_profile_samples",
+			Help:    "The number of samples per profile after downsampling",
+			Buckets: prometheus.ExponentialBuckets(32, 2, 15),
+		}, []string{"interval"})
+)
+
+type profilesWriter struct {
+	*parquet.GenericWriter[*schemav1.Profile]
+	file *os.File
+
+	buf []parquet.Row
+}
+
+func (p *profilesWriter) WriteRow(r parquet.Row) error {
+	p.buf[0] = r
+	_, err := p.GenericWriter.WriteRows(p.buf)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func newProfilesWriter(path string, i interval, aggregation string) (*profilesWriter, error) {
+	profilePath := filepath.Join(path, fmt.Sprintf("profiles_%s_%s", i.shortName, aggregation)+block.ParquetSuffix)
+	profileFile, err := os.OpenFile(profilePath, os.O_RDWR|os.O_CREATE|os.O_EXCL, 0o644)
+	if err != nil {
+		return nil, err
+	}
+	return &profilesWriter{
+		GenericWriter: newParquetProfileWriter(profileFile, parquet.MaxRowsPerRowGroup(100_000)),
+		file:          profileFile,
+		buf:           make([]parquet.Row, 1),
+	}, nil
+}
+
+func newParquetProfileWriter(writer io.Writer, options ...parquet.WriterOption) *parquet.GenericWriter[*schemav1.Profile] {
+	options = append(options, parquet.PageBufferSize(3*1024*1024))
+	options = append(options, parquet.CreatedBy("github.com/grafana/pyroscope/", build.Version, build.Revision))
+	options = append(options, schemav1.DownsampledProfilesSchema)
+	return parquet.NewGenericWriter[*schemav1.Profile](
+		writer, options...,
+	)
+}
+
+type aggregationType struct {
+	fn   func(a, b int64) int64
+	name string
+}
+
+var (
+	SumAggregation = aggregationType{
+		fn: func(a, b int64) int64 {
+			return a + b
+		},
+		name: "sum",
+	}
+)
```

Comment:
It seems the aggregation strategy assumes sample-level aggregation, which I'm not sure is what we actually need.

For example, consider two profiles:
```
a b c  c     
2 3 14 6

c d e f
4 5 3 1
```

My understanding is that their average is:
```
a b c  d e  
1 1 12 2 1
```

TBH, I can't see how such aggregation can be implemented using a `func(a, b int64) int64` function.

---

I'd propose to have something more context-aware: 

```go
type samples struct {
  ts     int64
  dur    int64
  total  int64
  ids    []parquet.Value
  values []parquet.Value
  // and all the fields we need
}

// or downsampling strategy
type aggregator interface {
   add(samples)
   build() samples
}
```

Reply:
Hm, actually, given that `avg` is not associative, we probably should use `sum` and keep track of the count (of profiles contributes to particular row), and only divide once, in the very end, after we get sum of all the profiles

Reply:
Added one simple way to support `avg` in [5a3f3e1](https://github.com/grafana/pyroscope/pull/2880/commits/5a3f3e16892e9a73fbfbc204a3a9ceb6f8192ffb). It would be easy to disable if we only want a `sum` for now.

Reply:
> I kept it simple because we can do many simple aggregations such as sum, min, max, first, last and count this way.

Yep, and your solution looks good!

My point is that we should aggregate profiles, not samples. For example, if we were to implement `max` aggregation, we'd need to pick the max profile for a given aggregation period (resolution) and not aggregate max stack trace values into a single profile.

But to be honest, I barely can imagine we will ever need anything besides `sum` and `avg`. Moreover, I don't think we benefit from a dedicated table/aggregation for `avg` as we can't actually use it because `avg` is not associative (sum/merge of two `avg` profiles is not _necessarily_ equal to the average of source profiles).

Therefore, I propose to implement only `sum` for now. Later, when we add `avg` aggregation for profiles to the querier, we'll add `count` column to the schema for downsampled data. Average profile is to be queried as a regular one, summing all the samples and counting source profiles; the average profile only emerges in the very end by dividing the total sum by the total count. Tricky point here is that a source profile can be split into multiple rows (due to pprof labels), but this is out of scope of the current work